File: app/src/main/java/com/example/mobile6/ui/base/BaseAdapter.java
package com.example.mobile6.ui.base;

import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

public abstract class BaseAdapter<T, VH extends RecyclerView.ViewHolder>
        extends ListAdapter<T, VH> {

    protected BaseAdapter(@NonNull DiffUtil.ItemCallback<T> diffCallback) {
        super(diffCallback);
    }

    @NonNull
    @Override
    public abstract VH onCreateViewHolder(@NonNull ViewGroup parent, int viewType);

    @Override
    public abstract void onBindViewHolder(@NonNull VH holder, int position);

    @Override
    public int getItemViewType(int position) {
        return super.getItemViewType(position);
    }
}

File: app/src/main/java/com/example/mobile6/presentation/ui/base/BaseAdapter.kt
package com.example.mobile6.presentation.ui.base

import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView

abstract class BaseAdapter<T, VH : RecyclerView.ViewHolder>(
    diffCallback: DiffUtil.ItemCallback<T>
) : ListAdapter<T, VH>(diffCallback) {
    companion object {
        inline fun <reified T : Any> simpleDiffCallback(
            crossinline areItemsTheSame: (T, T) -> Boolean = { old, new -> old == new },
            crossinline areContentsTheSame: (T, T) -> Boolean = { old, new -> old == new }
        ): DiffUtil.ItemCallback<T> = object : DiffUtil.ItemCallback<T>() {
            override fun areItemsTheSame(oldItem: T, newItem: T): Boolean =
                areItemsTheSame(oldItem, newItem)

            override fun areContentsTheSame(oldItem: T, newItem: T): Boolean =
                areContentsTheSame(oldItem, newItem)
        }
    }
}